src/hugr/views/sibling.rs

```diff
@@ -0,0 +1,572 @@
+///
+/// The ordering of the nodes in the subgraph is irrelevant to define the convex
+/// subgraph, but it determines the ordering of the boundary signature.
+#[derive(Clone, Debug)]
+pub struct SiblingSubgraph<'g, Base> {
+    /// The underlying Hugr.
+    base: &'g Base,
+    /// The nodes of the induced subgraph.
+    nodes: Vec<Node>,
+}
+
+impl<'g, Base: HugrInternals> SiblingSubgraph<'g, Base> {
+    /// A sibling subgraph from a HUGR.
+    ///
+    /// The subgraph is given by the sibling graph of the root. If you wish to
+    /// create a subgraph from another root, wrap the argument `region` in a
+    /// [`super::SiblingGraph`].
+    ///
+    /// This panics if the sibling graph is empty.
+    pub fn from_sibling_graph(sibling_graph: &'g Base) -> Self
+    where
+        Base: HugrView,
+    {
+        let root = sibling_graph.root();
+        let nodes = sibling_graph.children(root).collect_vec();
+        if nodes.is_empty() {
+            panic!("Empty graph");
+        }
+        Self {
+            base: sibling_graph,
+            nodes,
+        }
+    }
+
+    /// A sibling subgraph from a [`crate::ops::OpTag::DataflowParent`]-rooted HUGR.
+    ///
+    /// The subgraph is given by the nodes between the input and output
+    /// children nodes of the parent node. If you wish to create a subgraph
+    /// from another root, wrap the `region` argument in a [`super::SiblingGraph`].
+    ///
+    /// This panics if the DFG graph is empty.
+    pub fn from_dataflow_graph(dfg_graph: &'g Base) -> Self
+    where
+        Base: HugrView,
+    {
+        let parent = dfg_graph.root();
+        let nodes = dfg_graph.children(parent).skip(2).collect_vec();
+        if nodes.is_empty() {
+            panic!("Empty DFG graph");
+        }
+        Self {
+            base: dfg_graph,
+            nodes,
+        }
+    }
+
+    /// Create a new sibling subgraph from some boundary edges.
+    ///
+    /// Any sibling subgraph can be defined using two sets of boundary edges
+    /// $B_I$ and $B_O$, the incoming and outgoing boundary edges respectively.
+    /// Intuitively, the sibling subgraph is all the edges and nodes between
+    /// an edge of $B_I$ and an edge of $B_O$.
+    ///
+    /// The `incoming` and `outgoing` arguments give $B_I$ and $B_O$ respectively.
+    /// They can be either source or target ports. We currently assume that if
+    /// the source port of an outgoing boundary edge is linked to multiple
+    /// target ports, then all edges from the same source port are outgoing
+    /// boundary edges.
+    ///
+    /// More fomally, the sibling subgraph defined by $B_I$ and $B_O$ is the
+    /// graph given by the connected components of the graph with edges
+    /// $E\B_I\B_O$ which contain at least one node that is either
+    ///  - the target of an incoming boundary edge, or
+    ///  - the source of an outgoing boundary edge.
+    ///
+    /// A subgraph is well-formed if every edge in B into the subgraph is an
```

Comment:
This paragraph needs rephrasing given that incoming/outgoing are already separated. Something like "there are no incoming edges other than those in `incoming` and similarly for `outgoing`....no, hang on, that's guaranteed. IOW I'm not sure what the well-formedness criterion is here...

Reply:
I've gone with 

> A subgraph is well-formed if for every edge in the HUGR
>  - it is in $B_I$ if and only if it has a source outside of the subgraph and a target inside of it, and
 > - it is in $B_O$ if and only if it has a source inside of the subgraph and a target outside of it.

Reply:
There may be shorter ways but that seems good to me!